consensus/src/quorum_store/quorum_store_wrapper.rs

```diff
@@ -93,20 +97,40 @@ impl QuorumStoreWrapper {
         // TODO: size and unwrap or not?
         let pulled_txns = self
             .mempool_proxy
-            .pull_internal(MAX_SUB_BATCH_SIZE as u64, exclude_txns)
+            .pull_internal(MAX_FRAGMENT_SIZE as u64, exclude_txns)
             .await
             .unwrap();
-        self.batch_in_progress
-            .extend(pulled_txns.iter().map(|txn| TransactionSummary {
-                sender: txn.sender(),
-                sequence_number: txn.sequence_number(),
-            }));
 
-        // TODO: that is a bug! QuorumStore counts in bytes - need to keep track of it.
-        // TODO: also some timer if there are not enough txns
-        if self.batch_in_progress.len() <= self.max_batch_size - MAX_SUB_BATCH_SIZE {
+        let mut end_batch = false;
+        let mut txns_data = Vec::new();
+
+        // TODO: pass TxnData to QuorumStore to save extra serialization.
+        let mut pulled_txns_cloned = pulled_txns.clone();
+        for txn in pulled_txns {
+            let bytes = to_bytes(&txn).unwrap();
+            if self.bytes_in_progress + bytes.len() > self.max_batch_size {
+                end_batch = true;
+                break;
+            } else {
+                self.batch_in_progress.push(TransactionSummary {
+                    sender: txn.sender(),
+                    sequence_number: txn.sequence_number(),
+                });
+                self.bytes_in_progress = self.bytes_in_progress + bytes.len();
+                let mut hasher = DefaultHasher::new(b"TxnData");
+                hasher.update(&bytes);
+                txns_data.push(TxnData {
+                    txn_bytes: bytes,
+                    hash: hasher.finish(),
```

Comment:
Actually, if we're writing a hash per transaction, isn't this significantly more expensive than the approach of hash per QuorumStoreBatch? The size and also the setup cost of creating a new hasher per txn might be significant.

Reply:
Not sure. We are hashing the same amount of data. Do you think creating an hasher is a lot of overhead? Anyway, the hash is only for the addressable storge Rati wants to implement, so if we don't do we don't need the hashes.